keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,556 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, `kube-apiserver` allows one to configure the authorization chain using a set of command line flags of the format `--authorization-*`.
+And cluster administrators can only enable one webhook as a part of the authorization chain using the `--authorization-modes` flag.
+This limits admins from creating authorization chains with multiple webhooks that validate requests in a certain order.
+This proposal makes the case for a more structured config to define the authorization chain for `kube-apiserver` and allowing multiple webhooks with well-defined parameters and enabling fine grained control like explicit Deny authorizer and more definitive `SubjectAccessReviews`.
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization webhook in the API Server handler chain.
+They also need to be able to declaratively configure the authorizer chain using a configuration file.
+
+It should also be easy to say when to Deny requests, for example, when a webhook is unreachable.
+
+### Goals
+
+- Define a configuration file format for configuring Kubernetes API Server Authorization chain.
+- Allow ordered definition of authorization modes.
+- Allow definition of multiple webhooks in the authorization chain.
+- Enable user to define the policy when a webhook can't be reached due to network errors etc.
+- Enable more defnitive results for a `SubjectAccessReview` by allowing config to distinguish webhooks which might deny access.
+
+### Non-Goals
+
+- Allow resource/user based pre-filtering of webhooks to prevent unnecessary invocations.
+- Hot reload of the configuration.
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    server: https://<path-to-webserver>
+    authentication:
+        certificateAuthority: </path/to/ca.pem> # either this or certificateAuthorityData
+        clientCertificate: </path/to/cert.pem> # either this or clientCertificateData
+        clientKey: </path/to/key.pem> # either this or clientKeyData
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: Deny
+- type: Node
```

Comment:
if we allow something to come in front of system:masters (openshift wanted this), we'll need to have a type for `SystemPrivileged` to allow it to be ordered in here.

Reply:
A webhook ordered before the other auth modes should be able to deny a privileged user like `system:masters` access.
Would that not resolve this use case?

Reply:
No because today `system:masters` is always evaluated first.